File: day_02.rb
class PasswordDefinition
  attr_accessor :password, :min, :max, :pos_a, :pos_b, :letter

  def initialize(raw_definition)
    rules, password = raw_definition.split(':')
    rules = rules.split(/[\s,-]/)
    @password = password.strip
    @min = Integer(rules[0])
    @max = Integer(rules[1])
    @pos_a = @min - 1
    @pos_b = @max - 1
    @letter = rules[2]
  end

  def part_1_valid?
    count = @password.count(@letter)
    @min <= count && count <= @max
  end

  def part_2_valid?
    positions = (0...@password.length).find_all { |i| @password[i, 1] == @letter }
    (!positions.include?(@pos_a) && positions.include?(@pos_b)) || (positions.include?(@pos_a) && !positions.include?(@pos_b))
  end
end

text = File.open('day_02_input.txt').read
part_1_count = part_2_count = 0
text.each_line do |line|
  definition = PasswordDefinition.new(line)
  part_1_count += 1 if definition.part_1_valid?
  part_2_count += 1 if definition.part_2_valid?
end
puts "Part 1: #{part_1_count}"
puts "Part 2: #{part_2_count}"

File: day_02.py
with open('day_02_input.txt', 'r') as f:
    raw = f.readlines()
data = list()
part_1_valid_count = part_2_valid_count = 0
for line in raw:
    info = line.split(':')
    limits = info[0].split('-')
    min = int(limits[0])
    max = int(limits[1].split(' ')[0])
    first_position = min - 1
    second_position = max - 1
    letter = limits[1].split(' ')[1]
    password = info[1].strip()
    letter_count = password.count(letter)
    if min <= letter_count <= max:
        part_1_valid_count += 1
    positions = [pos for pos, char in enumerate(password) if char == letter]
    if (first_position in positions and second_position not in positions) or \
       (first_position not in positions and second_position in positions):
        part_2_valid_count += 1

print("Part 1: Valid passwords: {}".format(part_1_valid_count))
print("Part 2: Valid passwords: {}".format(part_2_valid_count))
